Feature/executor example (#39)
jannisspeer/snakemake-executor-plugin-htcondor: examples/basic-workflow/README.md

# Basic Workflow and Resources

A simple example demonstrating basic HTCondor job submission with resource requests.

> Find out more about plugin's resources and configuration under ["Basic"](https://github.com/htcondor/snakemake-executor-plugin-htcondor/blob/main/README.md) from the official executor's GitHub repository.

Reviewer: I took a quick look at the linked README. 
One thing I recommend addressing there is to add a about the "partial shared fs" option that warns that the user should check with the administrator of said FS about limitations (i.e., number of items, direct reads/writes, etc.).

## How This Example Works

> This example runs a simple Snakemake's rule called `process` that transforms the the input contents (i.e: `sample1.txt` and `sample2.txt`) to outputs by replacing the string `input` in the each `.txt` with the string `output`.

Reviewer: ```suggestion
This example runs a simple Snakemake's rule called `process` that transforms the input contents (i.e., `sample1.txt` and `sample2.txt`) to outputs by replacing the string `input` in each `.txt` file with the string `output`.
```

### Snakefile's Structure

- Line 1: `SAMPLES = ["sample1", "sample2"]` defines the set of sample names the workflow should run on.
- Line 3: `rule all` specifies target rule, which is what our final goals (output files) are for running the entire workflow
- Line 7: `rule process` does the actual processing of our data, specifying the input files, output files, and the shell command to generate the outputs from the inputs. This is the standard specifications for a rule in Snakemake.

### Log files

> **HTCondor**

Reviewer: ```suggestion
**HTCondor**

```

In `.snakemake`, you can find `htcondor` directory that contains `.err`, `.log`, and `.out` files. Each rule in a Snakefile generally produces a set of files (`.err`, `.log`, `.out`), with some exceptions such as that of grouped job.

> - `.log` records significant events that occur during the lifetime of all jobs within a cluster. This is important in helping us understanding what happened to a job and diagnosing if any issues occurred while the job was running/trying to run.

Reviewer: ```suggestion
- `.log` records significant events that occur during the lifetime of all jobs within a cluster. This is important to help us understand what happened to a job and to diagnose if any issues occurred while the job was running/trying to run.
```

- `.err` is the standard error, which typically contains the error messages.

> - `.out` is the standard output. Instead of printing to the console, all the outputs that would have printed in the console normally would be redirected here instead.

Reviewer: ```suggestion
- `.out` is the standard output, which records "normal" messages.
- `.err` is the standard error. which records error messages or additional information.

> All the outputs that would normally have been printed in the console are instead redirected to the appropriate `.out` or `.err` file.
```

> **Snakemake**

Reviewer: ```suggestion
**Snakemake**

```

# Basic Workflow and Resources

A simple example demonstrating basic HTCondor job submission with resource requests.

Find out more about plugin's resources and configuration under ["Basic"](https://github.com/htcondor/snakemake-executor-plugin-htcondor/blob/main/README.md) from the official executor's GitHub repository.

> ## How This Example Works

Reviewer 2: Inside this section, can you enumerate all the files in the folder and explain what their purpose is?

- Line 3: `rule all` specifies target rule, which is what our final goals (output files) are for running the entire workflow
- Line 7: `rule process` does the actual processing of our data, specifying the input files, output files, and the shell command to generate the outputs from the inputs. This is the standard specifications for a rule in Snakemake.

### Log files

**HTCondor**

In `.snakemake`, you can find `htcondor` directory that contains `.err`, `.log`, and `.out` files. Each rule in a Snakefile generally produces a set of files (`.err`, `.log`, `.out`), with some exceptions such as that of grouped job.

- `.log` records significant events that occur during the lifetime of all jobs within a cluster. This is important to help us understand what happened to a job and to diagnose if any issues occurred while the job was running/trying to run.
- `.out` is the standard output, which records "normal" messages.
- `.err` is the standard error. which records error messages or additional information.

All the outputs that would normally have been printed in the console are instead redirected to the appropriate `.out` or `.err` file.

**Snakemake**

> In addition to the HTCondor log, we also have the Snakemake's log file which can be found in `log` directory under `.snakemake`. The contents in the file is also printed to `Terminal`. This file contains the setups, job execution details and progress, warnings/errors, final status, and more. These are useful for understanding, troubleshooting, and inspecting the workflow that is being run.

Reviewer 2: Can you add another note after this about how the user can change the default logging dir? We should mention they can do this using either a command line flag, e.g. `--htcondor-jobdir logs/` or via the profile, e.g.:
```yaml
htcondor-jobdir: logs/
```

As someone with prior HTCondor experience, it feels strange to me having the logs buried in the hidden `.snakemake` directory. That being said, I think it's a fine default for the examples, since that's maybe what snakemake users are more familiar with.

When you do this, also remember to update the git ignore to include:
```
/examples/**/logs/
```

I think it's fine if this is the _only_ example you leave this comment for, since I assume most readers will start by reading the "basic workflow" example before moving onto something else.

### Plugin's Resources Used

Resources are set as **default-resources** in `htcondor_profile/config.yaml`

- `request_memory` — memory requested from HTCondor (default: 1GB)

> - `request_disk` — disk space requested from HTCondor (default: 4GB)

Reviewer 2: One of the benefits of looking through each job's `.log` file is that you can see how much of the requested resources the job actually used.

From running this example, I see we can slim these resource requests down. Can you update these to:
- 1MB of disk (big shrinkage!)
- 512MB of memory

These are useful for transferring files that aren't part of the rule's `input:`/`output:` directives (e.g., helper scripts, configuration files, logs, intermediate results).
Supports both string (comma-separated) and list formats. Wildcards (e.g., `{sample}`) are expanded for individual jobs, but **not** for grouped jobs since the resources are defined at the group level.
Files on shared filesystem prefixes are automatically excluded from transfer.

Example usage:

```python
rule process:
    input: "data/{sample}.txt"
    output: "results/{sample}.out"
    resources:
        htcondor_transfer_input_files="scripts/helpers.py,config/params.yaml",
        htcondor_transfer_output_files="logs/{sample}.log"
    script: "scripts/process.py"
```

## How This Example Works

Resources are set as **default-resources** in `profile/config.yaml`

- `request_memory` — memory requested from HTCondor (default: 1GB)
- `request_disk` — disk space requested from HTCondor (default: 4GB)
- `threads` — maps to HTCondor's `request_cpus` (default: 1)

These defaults apply to every rule in the workflow. For jobs with different

> requirements, you can override them per-rule using `resources:`

Author: WIP

### How to Run

```bash
snakemake --profile profile
```

### Expected Output

```

> results/sample1.txt

Author: WIP
